dynamic_programming/minimum_tickets_cost.py

```diff
@@ -0,0 +1,143 @@
+leetcode: https://leetcode.com/problems/minimum-cost-for-tickets/
+
+
+Minimum Cost For Tickets
+Dynamic Programming: up -> down.
+"""
+
+
+def mincost_tickets(days: list[int], costs: list[int]) -> int:
+    """
+    >>> mincost_tickets([1, 4, 6, 7, 8, 20], [2, 7, 15])
+    11
+
+    >>> mincost_tickets([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 30, 31],  [2, 7, 15])
+    17
+
+    >>> mincost_tickets([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 30, 31], [2, 90, 150])
+    24
+
+    >>> mincost_tickets([2], [2, 90, 150])
+    2
+
+    >>> mincost_tickets([], [2, 90, 150])
+    0
+
+    >>> mincost_tickets('hello', [2, 90, 150])
+    Traceback (most recent call last):
+     ...
+    ValueError: The parameter days should be a list
+
+    >>> mincost_tickets([], 'world')
+    Traceback (most recent call last):
+     ...
+    ValueError: The parameter costs should be a list
+
+    >>> mincost_tickets([0.25, 2, 3, 4, 5, 6, 7, 8, 9, 10, 30, 31], [2, 90, 150])
+    Traceback (most recent call last):
+     ...
+    ValueError: All days elements should be integer values
+
+    >>> mincost_tickets([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 30, 31], [2, 0.9, 150])
+    Traceback (most recent call last):
+     ...
+    ValueError: All costs elements should be integer values
+
+    >>> mincost_tickets([-1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 30, 31], [2, 90, 150])
+    Traceback (most recent call last):
+     ...
+    ValueError: All days elements should be greater than 0
+
+    >>> mincost_tickets([2, 367], [2, 90, 150])
+    Traceback (most recent call last):
+     ...
+    ValueError: All days elements should be less than 366
+
+    >>> mincost_tickets([2, 3, 4, 5, 6, 7, 8, 9, 10, 30, 31], [])
+    Traceback (most recent call last):
+     ...
+    ValueError: The lengths of costs should be equal 3
+
+    >>> mincost_tickets([], [])
+    Traceback (most recent call last):
+     ...
+    ValueError: The lengths of costs should be equal 3
+
+    >>> mincost_tickets([2, 3, 4, 5, 6, 7, 8, 9, 10, 30, 31], [1, 2, 3, 4])
+    Traceback (most recent call last):
+     ...
+    ValueError: The lengths of costs should be equal 3
+    """
+
+    # Validation
+    if not isinstance(days, list):
+        raise ValueError("The parameter days should be a list")
+
+    if not isinstance(costs, list):
+        raise ValueError("The parameter costs should be a list")
+
+    if any([not isinstance(day, int) for day in days]):
+        raise ValueError("All days elements should be integer values")
+
+    if any([not isinstance(cost, int) for cost in costs]):
+        raise ValueError("All costs elements should be integer values")
+
+    if len(costs) != 3:
+        raise ValueError("The lengths of costs should be equal 3")
+
+    if len(days) == 0:
+        return 0
+
+    if min(days) <= 0:
+        raise ValueError("All days elements should be greater than 0")
+
+    if max(days) >= 366:
+        raise ValueError("All days elements should be less than 366")
+
+    from functools import lru_cache
```

Comment:
Please import at a global scope (at the top of the file).

Reply:
fixed